File: src/main/cpp/weismarts/zmq/ZMsg.cpp
//
// Created by weismarts on 9/26/21.
//

#include "ZMsg.h"


ZMsg::ZMsg(){

};

void ZMsg::Add(char* msg)
{
    _multipart.addstr(std::string(msg));
}

void ZMsg::Add(void* msg, size_t len)
{
    _multipart.addmem(msg, len);
}

void ZMsg::Add(std::string msg)
{
    _multipart.addstr(msg);
}
void ZMsg::Add(zmq::message_t &msg)
{
    _multipart.add(zmq::message_t(msg.data(), msg.size()));
}
void ZMsg::AddAll(std::vector<std::string> msgs)
{
    for (int i = 0; i < msgs.size(); ++i) {
        const std::string msg =  msgs[i];//std::string ((char*)msgs[i].data(), msgs[i].size());
        _multipart.addstr(msg);
    }
}

ZMsg ZMsg::RecvMsg(ZMQSocket *socket, bool wait)
{
    ZMsg zMsg;
    std::vector<zmq::message_t> recv_msgs;
    const auto ret = zmq::recv_multipart(
            *socket->getPtr(), std::back_inserter(recv_msgs), wait?  zmq::recv_flags::none : zmq::recv_flags::dontwait);

    if(!ret) return zMsg;

    const size_t len =  recv_msgs.size();
    for (int i = 0; i < len; ++i) {
        zMsg.Add(recv_msgs[i].data(), recv_msgs[i].size());
    }

    return zMsg;
}

void ZMsg::Send(ZMQSocket *socket, int flags)
{
    _multipart.send(*socket->getPtr(), flags);
}
size_t ZMsg::Size()
{
    return _multipart.size();
}
std::string ZMsg::Peekstr(size_t index)
{
    return _multipart.peekstr(index);
}

std::string ZMsg::Str()
{
    return _multipart.str();
}

void *ZMsg::AtData(size_t n) {
    return _multipart.at(n).data();
}

void ZMsg::Destroy()
{
    _multipart.clear();
}

File: src/main/java/com/weismarts/librarys/ZMsg.java
package com.weismarts.librarys;

import java.nio.charset.Charset;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Iterator;

public class ZMsg {
    private ArrayDeque<ZFrame> _msgs;
    public ZMsg()
    {
        _msgs = new ArrayDeque<ZFrame>();
    }

    public void add(String msg)
    {
        _msgs.add(new ZFrame(msg));
    }

    public void add(byte[] msg)
    {
        _msgs.add(new ZFrame(msg));
    }

    public void destroy()
    {
        _msgs.clear();
    }
    public boolean send(ZMQSocket socket)
    {
        return send(socket, true);
    }
    public boolean send(ZMQSocket socket, boolean destroy)
    {
        if (_msgs.size() == 0) {
            return true;
        }

        boolean ret = true;
        Iterator<ZFrame> i = _msgs.iterator();
        while (i.hasNext()) {
            ZFrame f = i.next();
            send(socket.getID(), f ,(i.hasNext()) ? 2 : 0);
            if (!ret) {
                break;
            }
        }
        if (destroy) {
            destroy();
        }
        return ret;
    }
    //1 or 2 (1-none,2-more)
    public boolean sendMultipart(ZMQSocket socket)
    {
        Iterator<ZFrame> i = _msgs.iterator();
        ArrayList<ZFrame> msgs = new ArrayList<>();
        while (i.hasNext()) {
            ZFrame f = i.next();
            msgs.add(f);
        }
        return sendMultipart(socket.getID(),msgs, 2);
    }

    public ZFrame getLast()
    {
        return _msgs.peekLast();
    }


    static public ZMsg recvMsg(ZMQSocket socket)
    {
        return recvMsg(socket.getID());
    }

    private native boolean send(int socketID, ZFrame msg, int flags);//1 or 2 (1-none,2-more)

    private native boolean sendMultipart(int socketID, ArrayList<ZFrame> msgs, int flags);//1 or 2 (1-none,2-more)
    static public native ZMsg recvMsg(int socketID);


    // weismarts.zmq::message_t
    public class ZFrame
    {
        private byte[] _data;
        public ZFrame(String data)
        {

            if(data!=null)
            {
                _data = data.getBytes(Charset.forName("UTF-8"));
            }
        }

        public ZFrame(byte[] data)
        {
            if(data!=null)
            {
                _data = data;
            }
        }

        public byte[] getData()
        {
            return _data;
        }

        public int size()
        {
            if (_data!=null) {
                return _data.length;
            }
            else {
                return 0;
            }
        }

        public String toString()
        {
            return new String(_data, Charset.forName("UTF-8"));
        }
    }

}


